sdk/cosmos/azure-cosmos/azure/cosmos/aio/_cosmos_client_connection_async.py

```diff
@@ -3052,27 +3052,40 @@ def GetPartitionResolver(self, database_link: str) -> Optional[RangePartitionRes
     # Adds the partition key to options
     async def _AddPartitionKey(self, collection_link, document, options):
         collection_link = base.TrimBeginningAndEndingSlashes(collection_link)
+        new_options = dict(options)
+        # If we get an Empty or Undefined Partition key, refresh the container properties cache
+        # and try one more time.
+        for attempts in range(2):
```

Comment:
Why does this need to be in a loop?

Reply:
Never mind - looks like this was a comment in the sync version of this class.

Reply:
Doing this as a rule for every data plane operation will overload the gateway wouldn't it - if the data plane request fails due to the container cache staleness and then you refresh the container cache makes but not prior to executing a request IMO? @simorenoh, @kushagraThapar @bambriz @xinlian12

Reply:
I should note that the code you are referring to is outdated, in the latest changes the retry logic is all done in retry utils and so there is no changes here.